### PR TITLE
add support for lists, ref #80

### PR DESCRIPTION
**Pull request by gcanti**
This adds support for lists. 

Builds on my previous PR https://github.com/gcanti/tcomb-form-native/pull/172

It's a direct porting of the same `tcomb-form` feature, this means you can handle lists of strings, lists of structs, lists of lists, lists of unions of lists of structs... etc etc etc :)

The engine (the `List` component in `components.js`) should be ok, but I need some help with the default template: I put up a bare and ugly one, just for having something to show. 

I'm a web developer :) when it comes to mobile development and RN it's better having someone more experienced than me to jump in and change the code.

### CODE REVIEW

**Comment by alvaromb** [comment]
Hey @gcanti, this is awesome. Haven't had time to dig into this this weekend, please tell me what do you need to bring this into master, I'm currently trying to implement this into one of our projects.

File: lib/components.js [hunk 1]
```diff
@@ -510,14 +520,228 @@ class Struct extends Component {
...
+      }
+      if (!options.disableOrder) {
+        buttons.push({
+          type: 'move-up',
+          label: i18n.up,
+          click: this.moveUpItem.bind(this, i)
+        });
+      }
+      if (!options.disableOrder) {
+        buttons.push({
```
Comment (alvaromb) [inline lib/components.js]: Couldn't we move this `push` into the upper `if`? It's the same condition.

**Comment by alvaromb** [comment]
I'm curious, how would you define a fixed list of different items? For example, you can toggle (`true`/`false`) a list of items with different name. Should that be a union?

File: lib/components.js (hunk 1 above)
Comment (gcanti) [inline lib/components.js]: Yep, I'll fix that

**Comment by gcanti** [comment]
> how would you define a fixed list of different items?

@alvaromb not sure I understand. Could you elaborate?

**Comment by gcanti** [comment]
> please tell me what do you need to bring this into master

The current default template for lists is rather ugly and not configurable (button styles, columns width, etc..)

**Comment by alvaromb** [comment]
The actual use case I'm developing is a filter by type form. Let's say we have a bunch of items that every one of it is related to a type. We have 4 types in total. I display the 4 types in the form and I can select/deselect each one.

**Comment by alvaromb** [comment]
> The current default template for lists is rather ugly and not configurable (button styles, columns width, etc..)

Ok! That one goes for me. I'll do that later today. Do you want me to push directly into this branch?

**Comment by gcanti** [comment]
> Do you want me to push directly into this branch?

I think it would be better to send PRs for this branch

> The actual use case I'm developing is a filter by type form

At first glance seems a struct of unions:

```js
const Toggle = t.struct({
  toggle: t.Boolean
})

const Type1 = Toggle.extend({
  name1: t.String
})

const Type2 = Toggle.extend({
  name2: t.String
})

const Type3 = Toggle.extend({
  name3: t.String
})

const Type4 = Toggle.extend({
  name4: t.String
})

const U1 = t.union([Toggle, Type1])
U1.dispatch = x => x && x.toggle ? Type1 : Toggle

const U2 = t.union([Toggle, Type2])
U2.dispatch = x => x && x.toggle ? Type2 : Toggle

const U3 = t.union([Toggle, Type3])
U3.dispatch = x => x && x.toggle ? Type3 : Toggle

const U4 = t.union([Toggle, Type4])
U4.dispatch = x => x && x.toggle ? Type4 : Toggle

const Type = t.struct({
  u1: U1,
  u2: U2,
  u3: U3,
  u4: U4
})
```

**Comment by alvaromb** [comment]
I was thinking about defining a struct for every type and then simply performing an union.

```js
const Type1 = t.struct({
  type1: t.Bool,
})
const Type2 = t.struct({
  type2: t.Bool,
})
const Type3 = t.struct({
  type3: t.Bool,
})
const Type4 = t.struct({
  type4: t.Bool,
})
const Filters = t.union([Type1, Type2, Type3, Type4])
```

Expected output should be something like: `({type1: true, type2: false, type3: false, type4: false})` & other possible combinations. I'm on the right path?

**Comment by gcanti** [comment]
> Expected output should be something like: ({type1: true, type2: false, type3: false, type4: false})

That is not a union (sum type) it's a struct (product type), the expected output of your `Filter` union would be:

```
{type1: true} or {type2: false} or {type3: false} or {type4: false}
```

You need a sentinel value in order to enable / disable the filter (hence the toggle boolean).

The expected output of my `Type` would be:

```
{
  u1: { toggle: false } or { toggle: true, name1: 'blah' },
  u2: { toggle: false } or { toggle: true, name2: 'blah' },
  ...
  u4: { toggle: false } or { toggle: true, name4: 'blah' }
}
```

**Comment by alvaromb** [comment]
Oh I see!

So, actually a union will return one of its types. With a list:

```js
const Type = t.struct({
  enabled: t.Bool,
  typeName: t.String,
})
const Filters = t.list(Type)
```

I could do what I need, but I think that in the end it's easier to define a struct with `t.Bool`'s, its simpler. Anyway, I'll submit some changes for the list components later today.

**Comment by alvaromb** [comment]
@gcanti I tweaked a bit the stylesheet for the list components

screen shot 2016-05-30 at 17 54 54

However, I would like to have some consensus about the changes you want to implement into these components. I have some ideas in mind, like rendering the list with a `ListView`, support cell swipe to delete, drag for reordering, but I prefer to first have a simpler version and then improve it with your suggestions.

**Comment by gcanti** [comment]
> I have some ideas in mind, like rendering the list with a ListView, support cell swipe to delete, drag for reordering, but I prefer to first have a simpler version and then improve it with your suggestions

Awesome, great ideas!

As a reminder, make sure that the default template will be cross platform and that it will be usable even if you have a list of structs instead of a list of strings ;-)

**Comment by JacoBezuidenhout** [comment]
Hey All... Any progress on this... This is a make or break feature!

**Comment by gcanti** [comment]
Hi @alvaromb,

Any progress? Otherwise we could also release the core `List` component in `component.js` and throw in the `list` template saying that a custom template must be implemented until a default one is ready. This way people can start to play with the new features and we would get some feedback (and hopefully some help).

**Comment by alvaromb** [comment]
Sorry, I'm overwhelmed right now. I want to finish the new collapsed date component API in the docs this week but I'm going to be unable to finish this task this month.

**Comment by alvaromb** [comment]
@gcanti however, I can find some time to apply the style changes I posted in the screenshot. Does that work for this release?

**Comment by gcanti** [comment]
@alvaromb would be great, thanks a lot

**Comment by gcanti** [comment]
Hi @alvaromb,

I'm going to merge and release this PR as it is, we can later refine the template.

**Pull request merged**